compiler/src/iree/compiler/Preprocessing/Passes.cpp

```
@@ -64,35 +64,56 @@ void extendWithTextPipeline(OpPassManager &passManager,
 
 } // namespace
 
-void buildPreprocessingPassPipeline(
+/// Adds passes to `passManager` based on command line options.
+/// Returns `true` if passes were added, `false` otherwise.
+static bool buildPreprocessingPassPipelineFromCommandLine(
     OpPassManager &passManager,
-    const PreprocessingOptions &preprocessingOptions,
-    PipelineExtensions *pipelineExtensions) {
+    const PreprocessingOptions &preprocessingOptions) {
   auto pipelineStr = preprocessingOptions.preprocessingPassPipeline;
+  // First preference is explicit pass pipeline.
   if (!pipelineStr.empty()) {
     extendWithTextPipeline(passManager, pipelineStr);
+    return true;
   }
-
-  if (pipelineExtensions) {
-    pipelineExtensions->extendPreprocessingPassPipeline(passManager);
-  }
-
-  if (!preprocessingOptions.preprocessingTransformSpecFilename.empty()) {
+  // Second preference is for transform spec file as a preprocessing recipe.
+  else if (!preprocessingOptions.preprocessingTransformSpecFilename.empty()) {
     Preprocessing::InterpreterPassOptions interpreterOptions;
     interpreterOptions.transformSpecPath =
         preprocessingOptions.preprocessingTransformSpecFilename;
     passManager.addPass(
         Preprocessing::createInterpreterPass(interpreterOptions));
+    return true;
   }
-
-  if (!preprocessingOptions.preprocessingPDLSpecFilename.empty()) {
+  // Third preference is for PDL spec file as a preprocessing recipe.
+  else if (!preprocessingOptions.preprocessingPDLSpecFilename.empty()) {
```

Do we really need to go to the effort to make these different options mutually exclusive? I get how using them together is unadvisible, but I think having one silently skip is more surprising.

I do like the new clarification on the flag docs though, lets definitely keep that.

---

After reconsidering it, I agree with you. I removed the "only one way" option. The comments now specify the order in which different options are exercised, but they arent exclusive.

---

Doesn't the `if` -> `else if` change make it exclusive?

---

Ah. Good catch. Will fix that.